### PR TITLE
Fixed fam_copy unit test fails as per #141

### LINKED ISSUE #141: fam_copy unit Test fails & memserver crashes on verbs provider
srun: job 137484 has been allocated resources
Content of source dataitem : Test message
Copy 0 : Copying first 1 characters
Copy 1 : Copying first 2 characters
Copy 2 : Copying first 3 characters
Copy 3 : Copying first 4 characters
Copy 4 : Copying first 5 characters
Copy 5 : Copying first 6 characters
Copy 6 : Copying first 7 characters
Copy 7 : Copying first 8 characters
Copy 8 : Copying first 9 characters
Copy 9 : Copying first 10 characters
Copy 10 : Copying first 11 characters
Copy 11 : Copying first 12 characters
waiting for Copy : 11
Exception caught
Error msg: channel is in state TRANSIENT_FAILURE
Error: 11
waiting for Copy : 10
Exception caught
Error msg: channel is in state TRANSIENT_FAILURE
Error: 11
waiting for Copy : 9
Exception caught
Error msg: channel is in state TRANSIENT_FAILURE
Error: 11
waiting for Copy : 8
Exception caught
Error msg: channel is in state TRANSIENT_FAILURE
Error: 11
waiting for Copy : 7
Exception caught
Error msg: channel is in state TRANSIENT_FAILURE
Error: 11
waiting for Copy : 6
Exception caught
Error msg: channel is in state TRANSIENT_FAILURE
Error: 11
waiting for Copy : 5
Exception caught
Error msg: channel is in state TRANSIENT_FAILURE
Error: 11
waiting for Copy : 4
Exception caught
Error msg: channel is in state TRANSIENT_FAILURE
Error: 11
waiting for Copy : 3
Exception caught
Error msg: channel is in state TRANSIENT_FAILURE
Error: 11
waiting for Copy : 2
Exception caught
Error msg: channel is in state TRANSIENT_FAILURE
Error: 11
waiting for Copy : 1
Exception caught
Error msg: Connect Failed
Error: 11
waiting for Copy : 0
Exception caught
Error msg: Socket closed
Error: 11
srun: error: xxxx3: task 0: Segmentation fault (core dumped)
srun: error: xxxx5: task 0: Segmentation fault (core dumped)
srun: error: xxxx7: task 0: Segmentation fault (core dumped)

### PR DESCRIPTION
**Pull request by rajakrishi**
Fixed #141

### CODE REVIEW

**Comment by rajakrishi** [comment]
Tested on verbs provider
srun: job 137817 has been allocated resources
Content of source dataitem : Test message
Copy 0 : Copying first 1 characters
Copy 1 : Copying first 2 characters
Copy 2 : Copying first 3 characters
Copy 3 : Copying first 4 characters
Copy 4 : Copying first 5 characters
Copy 5 : Copying first 6 characters
Copy 6 : Copying first 7 characters
Copy 7 : Copying first 8 characters
Copy 8 : Copying first 9 characters
Copy 9 : Copying first 10 characters
Copy 10 : Copying first 11 characters
Copy 11 : Copying first 12 characters
waiting for Copy : 11
waiting for Copy : 10
waiting for Copy : 9
waiting for Copy : 8
waiting for Copy : 7
waiting for Copy : 6
waiting for Copy : 5
waiting for Copy : 4
waiting for Copy : 3
waiting for Copy : 2
waiting for Copy : 1
waiting for Copy : 0
Copy 0 : T : CORRECT
Copy 1 : Te : CORRECT
Copy 2 : Tes : CORRECT
Copy 3 : Test : CORRECT
Copy 4 : Test  : CORRECT
Copy 5 : Test m : CORRECT
Copy 6 : Test me : CORRECT
Copy 7 : Test mes : CORRECT
Copy 8 : Test mess : CORRECT
Copy 9 : Test messa : CORRECT
Copy 10 : Test messag : CORRECT
Copy 11 : Test message : CORRECT
fam finalize successful
rishi@xxxx3:/nfs/rishi/OpenFAM/build$

File: src/cis/fam_cis.h [hunk 1]
```diff
@@ -258,7 +259,7 @@ class Fam_CIS {
      * @param gid - gid of user
      **/
     virtual void *copy(uint64_t srcRegionId, uint64_t srcOffset,
-                       uint64_t srcCopyStart, uint64_t srcKey,
+                       uint64_t srcCopyStart, uint64_t srcKey, uint64_t srcBaseAddr,
```
Comment (mashoodk) [inline src/cis/fam_cis.h:262]: Please run git-clang-format

**Review by mashoodk — CHANGES_REQUESTED** [review]
See if we should validate the copy output in "fam_copy_test.cpp". In fam_copy_test.cpp we are comparing the strings (original vs final), but we are not throwing any error.

**Comment by rajakrishi** [comment]
> See if we should validate the copy output in "fam_copy_test.cpp". In fam_copy_test.cpp we are comparing the strings (original vs final), but we are not throwing any error.

We can come out of loop with break once we fall in Wrong (aka else scope). No need for further compare string if one string fails. Let me know if that is fine with you or you were suggesting to throw exception if one char does not matches ?

**Review by BA-Ramya — COMMENTED** [review]
In file src/memory_service/fam_memory_service.h, variable name is srcBaseAdd.
May be we can change that to srcBaseAddr to maintain uniformity.

In file test/unit-test/fam-api/fam_copy_test.cpp 
change MESSAGE_SIZE to NUM_DATAITEMS as the name is misleading.

**Comment by rajakrishi** [comment]
> In file src/memory_service/fam_memory_service.h, variable name is srcBaseAdd. May be we can change that to srcBaseAddr to maintain uniformity.
> 
> In file test/unit-test/fam-api/fam_copy_test.cpp change MESSAGE_SIZE to NUM_DATAITEMS as the name is misleading.



> In file src/memory_service/fam_memory_service.h, variable name is srcBaseAdd. May be we can change that to srcBaseAddr to maintain uniformity.
> 
> In file test/unit-test/fam-api/fam_copy_test.cpp change MESSAGE_SIZE to NUM_DATAITEMS as the name is misleading.

Thanks Ramya. I have addressed the first input. We should maintain uniformity. I have checked in the code 
For second input, it is not NUM of Data item, it is MESSAGE size only to check and compare.

**Review by BA-Ramya — APPROVED** [review]
Review comments addressed

**Review by faizan-barmawer — APPROVED** [review]
Code changes look good.

Regards,
Faizan

**Review by mashoodk — APPROVED** [review]

**Pull request merged**